File: python/day_20b.py
#!/usr/bin/python3

import math
import copy
import builtins

class Tile:
    def __init__(self):
        self.used = False
        self.flipped = False
        self.id = -1
        self.rotation = 0
        self.neighbour_ids = set()
        self.image = list()
        self.image_str = list()
        self.edges = list()

    def createEdges(self):
        left = str()
        right = str()
        for line in self.image_str:
            left = left + line[0]
            right = right + line[len(line) - 1]
        self.edges.append(self.image_str[0])
        self.edges.append(right)
        self.edges.append(self.image_str[len(self.image_str) - 1])
        self.edges.append(left)

    def getRightEdge(self):
        return self.edges[1]

    def getBottomEdge(self):
        return self.edges[2]

    def Rotate(self):
        self.edges.insert(0, self.edges[3])
        self.edges.pop()
        self.edges[0] = self.edges[0][::-1]
        self.edges[2] = self.edges[2][::-1]
        self.rotation = self.rotation + 1
        self.rotation = self.rotation % 4

    def Flip(self):
        temp = copy.deepcopy(self.edges[0])
        self.edges[0] = copy.deepcopy(self.edges[1])
        self.edges[1] = copy.deepcopy(temp)
        temp = copy.deepcopy(self.edges[3])
        self.edges[3] = copy.deepcopy(self.edges[2])
        self.edges[2] = copy.deepcopy(temp)

        for i in range(0, len(self.edges)):
            self.edges[i] = self.edges[i][::-1]
        self.flipped = not self.flipped

    def applyRotationsAndFlipToImage(self):
        if self.flipped:
            new_image = copy.deepcopy(self.image)
            for i in range(0, len(self.image)):
                for j in range(0, len(self.image[i])):
                    new_image[len(self.image[i]) - j - 1][len(self.image) - i - 1] = self.image[i][j];
            self.image = copy.deepcopy(new_image)
        for r in range(0, self.rotation):
            new_image = copy.deepcopy(self.image)
            for i in range(0, len(self.image)):
                for j in range(0, len(self.image[i])):
                    new_image[j][len(self.image) - i - 1] = self.image[i][j];
            self.image = copy.deepcopy(new_image)

def RotateImage(image):
    new_image = copy.deepcopy(image)
    for i in range(0, len(image)):
        for j in range(0, len(image[i])):
            new_image[j][len(image) - i - 1] = image[i][j];
    return new_image

def FlipImage(image):
    new_image = copy.deepcopy(image)
    for i in range(0, len(image)):
        for j in range(0, len(image[i])):
            new_image[len(image[i]) - j - 1][len(image) - i - 1] = image[i][j];
    return new_image


def OrientTileToMatchOnSide(edge, cur_tile, edge_dir):
    for i in range(0, 4):
        if edge == cur_tile.edges[edge_dir]:
            return
        else:
            cur_tile.Rotate()
    cur_tile.Flip()
    for i in range(0, 4):
        if edge == cur_tile.edges[edge_dir]:
            return
        else:
            cur_tile.Rotate()

def CheckForPattern(complete_image, pattern):
    max_row_size = 0
    relevant_points = list()
    for i in range(0, len(pattern)):
      relevant_points.append([])
      for j in range(0, len(pattern[i])):
          max_row_size = max(max_row_size, len(pattern[i]))
          if pattern[i][j] == '#':
              relevant_points[i].append(j)
    count = 0;
    for i in range(0, len(complete_image) - len(relevant_points)):
        for j in range(0, len(complete_image[i]) - max_row_size):
            match = True
            for pi in range(0, len(relevant_points)):
                for pj in range(0, len(relevant_points[pi])):
                    if complete_image[i+pi][j+relevant_points[pi][pj]] != '#':
                        match = False
                        break
                if not match:
                    break
            if match:
                count = count + 1
    return count

def backtrack(puzzle_size, count, grid, edge_map, tiles):
    if count == len(tiles):
        return True

    row = math.floor(count / puzzle_size)
    col = int(count % puzzle_size)
    if col == 0:
        bottom_edge = tiles[grid[row - 1][col]].getBottomEdge()
        ids = edge_map[bottom_edge]
        for id in ids:
            if id == grid[row - 1][0] or tiles[id].used:
                continue;
            cur_tile = tiles[id]
            cur_tile.used = True
            grid[row][col] = cur_tile.id
            OrientTileToMatchOnSide(bottom_edge, cur_tile, 0)
            if backtrack(puzzle_size, count + 1, grid, edge_map, tiles):
                return True
            else:
                cur_tile.used = False
                grid[row][col] = -1;
    elif row == 0:
        right_edge = tiles[grid[row][col - 1]].getRightEdge()
        ids = edge_map[right_edge]
        for id in ids:
            if id == grid[row][col - 1] or tiles[id].used:
                continue;
            cur_tile = tiles[id]
            cur_tile.used = True
            grid[row][col] = cur_tile.id
            OrientTileToMatchOnSide(right_edge, cur_tile, 3)
            if backtrack(puzzle_size, count + 1, grid, edge_map, tiles):
                return True
            else:
                cur_tile.used = False
                grid[row][col] = -1;
    elif row != 0:
        bottom_edge = tiles[grid[row - 1][col]].getBottomEdge()
        ids = edge_map[bottom_edge]
        for id in ids:
            if id == grid[row - 1][col] or tiles[id].used:
                continue;
            cur_tile = tiles[id]
            cur_tile.used = True
            grid[row][col] = cur_tile.id
            OrientTileToMatchOnSide(bottom_edge, cur_tile, 0)
            right_edge = tiles[grid[row][col - 1]].getRightEdge()
            if cur_tile.edges[3] == right_edge and backtrack(puzzle_size, count + 1, grid, edge_map, tiles):
                return True
            else:
                cur_tile.used = False
                grid[row][col] = -1;
    return False

def main():
    f = open("../input/day_20_input")
    new_tile = Tile()
    tiles = dict()
    edge_map = dict()
    for line in f:
        line = line.strip('\n')
        if line == "":
            new_tile.createEdges()
            tiles[new_tile.id] = copy.deepcopy(new_tile)
            new_tile = Tile()
            continue
        elif line[0:4] == "Tile":
            new_tile.id = int(line[5:len(line) - 1])
        else:
            new_tile.image_str.append(line)
            new_tile.image.append([])
            for char in line:
                new_tile.image[len(new_tile.image) - 1].append(char)
    if new_tile.id != -1:
        new_tile.createEdges()
        tiles[new_tile.id] = new_tile

    for tile_id in tiles:
        tile = tiles[tile_id]
        for edge in tile.edges:
            if edge not in edge_map:
                edge_map[edge] = [tile_id]
                r_edge = edge[::-1]
                edge_map[r_edge] = [tile_id]
            else:
                if tile_id not in edge_map[edge]:
                    edge_map[edge].append(tile_id)
                    r_edge = edge[::-1]
                    edge_map[r_edge].append(tile_id)

    for edge in edge_map:
        ids = edge_map[edge]
        for i in range(0, len(ids)):
            for j in range(i + 1, len(ids)):
                if j not in tiles[ids[i]].neighbour_ids:
                    tiles[ids[i]].neighbour_ids.add(ids[j])
                if i not in tiles[ids[j]].neighbour_ids:
                    tiles[ids[j]].neighbour_ids.add(ids[i])

    corner_ids = list()
    for id in tiles:
        tile = tiles[id]
        if len(tile.neighbour_ids) == 2:
            corner_ids.append(id)

    found = False
    count = 1
    puzzle_size = int(math.sqrt(len(tiles)))
    grid = list()
    for i in range(0, puzzle_size):
        grid.append([-1] * puzzle_size)
    for corner_id in corner_ids:
        grid[0][0] = corner_id
        tiles[corner_id].used = True
        for i in range(0, 4):
            if backtrack(puzzle_size, count, grid, edge_map, tiles):
                found = True
                break
            tiles[corner_id].Rotate()
        if found:
            break
        tiles[corner_id].Flip();
        for i in range(0, 4):
            if backtrack(puzzle_size, count, grid, edge_map, tiles):
                found = True
                break
            tiles[corner_id].Rotate()
        if found:
            break
        else:
            tiles[corner_id].used = False
    complete_image = list()
    for i in range(0, puzzle_size):
        for j in range(0, puzzle_size):
            tiles[grid[i][j]].applyRotationsAndFlipToImage()
            tile_img = tiles[grid[i][j]].image
            for k in range(0, len(tile_img) - 2):
                # if (j == 0):
                #     complete_image.append([])
                if i * (len(tile_img) - 2) + k >= len(complete_image):
                    complete_image.append([])
                complete_image[i * (len(tile_img) - 2) + k].extend(tile_img[k+1][1:len(tile_img[k]) - 1]);
    pattern_temp = ["                  # ", "#    ##    ##    ###", " #  #  #  #  #  #   "]
    pattern = list()
    for i in range(0, len(pattern_temp)):
        pattern.append([])
        for char in pattern_temp[i]:
            pattern[i].append(char)

    monster_size = 0
    for row in pattern:
        for ele in row:
            if ele == '#':
                monster_size = monster_size + 1

    n_monsters = 0
    for i in range(0, 2):
        for j in range(0, 4):
            n_monsters = CheckForPattern(complete_image, pattern)
            if n_monsters > 0:
                break
            complete_image = RotateImage(complete_image)
        if n_monsters > 0:
            break
        complete_image = FlipImage(complete_image)

    roughness = 0
    for i in range(0, len(complete_image)):
        for j in range(0, len(complete_image[i])):
            if complete_image[i][j] == '#':
                roughness = roughness + 1
    roughness = roughness - (n_monsters * monster_size)
    print(roughness)
    return roughness

if __name__ == "__main__":
    main()

File: cpp/day_20b.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cmath>
#include <fstream>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

struct Tile {
  bool used = false;
  bool flipped = false;
  int id = -1;
  int rotation = 0;
  std::unordered_set<int> neighbour_ids;
  std::vector<std::string> image;
  std::vector<std::string> edges;  // always in order up, right, bottom, left
  // the sides are always read left to right and up to down

  void createEdges() {
    std::string left, right;
    for (int i = 0; i < image.size(); i++) {
      left += image[i][0];
      right += image[i][image[0].size() - 1];
    }
    edges.push_back(image.front());
    edges.push_back(right);
    edges.push_back(image.back());
    edges.push_back(left);
  }

  std::string& getRightEdge() { return edges[1]; }

  std::string& getBottomEdge() { return edges[2]; }

  void Rotate() {
    std::rotate(edges.rbegin(), edges.rbegin() + 1, edges.rend());
    std::reverse(std::begin(edges[0]), std::end(edges[0]));
    std::reverse(std::begin(edges[2]), std::end(edges[2]));
    rotation += 1;
    rotation %= 4;
  }

  void Flip() {
    std::swap(edges[0], edges[1]);
    std::swap(edges[2], edges[3]);
    for (auto& edge : edges) {
      std::reverse(std::begin(edge), std::end(edge));
    }
    flipped = !flipped;
  }

  void applyRotationsAndFlipToImage() {
    if (flipped) {
      auto new_image = image;
      for (int i = 0; i < image.size(); i++) {
        for (int j = 0; j < image[i].size(); j++) {
          new_image[image[i].size() - j - 1][image.size() - i - 1] =
              image[i][j];
        }
      }
      image = new_image;
    }
    for (int i = 0; i < rotation; i++) {
      auto new_image = image;
      for (int i = 0; i < image.size(); i++) {
        for (int j = 0; j < image[i].size(); j++) {
          new_image[j][image.size() - i - 1] = image[i][j];
        }
      }
      image = new_image;
    }
  }
};

void RotateImage(std::vector<std::string>& image) {
  auto new_image = image;
  for (int i = 0; i < image.size(); i++) {
    for (int j = 0; j < image[i].size(); j++) {
      new_image[j][image.size() - i - 1] = image[i][j];
    }
  }
  image = new_image;
}

void FlipImage(std::vector<std::string>& image) {
  auto new_image = image;
  for (int i = 0; i < image.size(); i++) {
    for (int j = 0; j < image[i].size(); j++) {
      new_image[image[i].size() - j - 1][image.size() - i - 1] = image[i][j];
    }
  }
  image = new_image;
}

// the tile has to match the edge passed in along a particular orientation, ie,
// edge_dir so if the edge passed in is a right edge of the prev tile, the
// edge_dir should be 3, for the right edge of the current tile 0 = up, 1 =
// right, 2 = left and 3 = right
void OrientTileToMatchOnSide(const std::string& edge, Tile& cur_tile,
                             const int edge_dir) {
  for (int i = 0; i < 4; ++i) {
    if (edge == cur_tile.edges[edge_dir]) {
      return;
    } else {
      cur_tile.Rotate();
    }
  }
  cur_tile.Flip();
  for (int i = 0; i < 4; ++i) {
    if (edge == cur_tile.edges[edge_dir]) {
      return;
    } else {
      cur_tile.Rotate();
    }
  }
}

bool backtrack(const int puzzle_size, const int count,
               std::vector<std::vector<int>>& grid,
               std::unordered_map<std::string, std::vector<int>>& edge_map,
               std::unordered_map<int, Tile>& tiles) {
  if (count == tiles.size()) return true;
  const int row = count / puzzle_size;
  const int col = count % puzzle_size;
  if (col == 0) {  // row and col is never 0 as set the tile for 0,0 before
                   // calling backtrack
    const auto& bottom_edge = tiles[grid[row - 1][col]].getBottomEdge();
    const auto& ids = edge_map[bottom_edge];
    for (const auto id : ids) {
      if (id == grid[row - 1][0] || tiles[id].used) continue;
      auto& cur_tile = tiles[id];
      cur_tile.used = true;
      grid[row][col] = cur_tile.id;
      OrientTileToMatchOnSide(bottom_edge, cur_tile, 0);
      if (backtrack(puzzle_size, count + 1, grid, edge_map, tiles)) {
        return true;
      } else {
        cur_tile.used = false;
        grid[row][col] = -1;
      }
    }
  } else if (row == 0) {
    const auto& right_edge = tiles[grid[row][col - 1]].getRightEdge();
    const auto& ids = edge_map[right_edge];
    for (const auto id : ids) {
      if (id == grid[row][col - 1] || tiles[id].used) continue;
      auto& cur_tile = tiles[id];
      cur_tile.used = true;
      grid[row][col] = cur_tile.id;
      OrientTileToMatchOnSide(right_edge, cur_tile, 3);
      if (backtrack(puzzle_size, count + 1, grid, edge_map, tiles)) {
        return true;
      } else {
        cur_tile.used = false;
        grid[row][col] = -1;
      }
    }
  } else if (row != 0) {
    const auto& bottom_edge = tiles[grid[row - 1][col]].getBottomEdge();
    const auto& ids = edge_map[bottom_edge];
    for (const auto id : ids) {
      if (id == grid[row - 1][col] || tiles[id].used) continue;
      auto& cur_tile = tiles[id];
      cur_tile.used = true;
      grid[row][col] = cur_tile.id;
      OrientTileToMatchOnSide(bottom_edge, cur_tile, 0);
      const auto& right_edge = tiles[grid[row][col - 1]].getRightEdge();
      if (cur_tile.edges[3] == right_edge &&
          backtrack(puzzle_size, count + 1, grid, edge_map, tiles)) {
        return true;
      } else {
        cur_tile.used = false;
        grid[row][col] = -1;
      }
    }
  }
  return false;
}

long long CheckForPattern(const std::vector<std::string>& complete_image,
                          const std::vector<std::string>& pattern) {
  std::size_t max_row_size = 0;
  std::vector<std::vector<int>> relevant_points(pattern.size());
  for (int i = 0; i < pattern.size(); i++) {
    for (int j = 0; j < pattern[i].size(); j++) {
      max_row_size = std::max(max_row_size, pattern[i].size());
      if (pattern[i][j] == '#') {
        relevant_points[i].push_back(j);
      }
    }
  }
  long long count = 0;
  for (int i = 0; i < complete_image.size() - relevant_points.size(); i++) {
    for (int j = 0; j < complete_image[i].size() - max_row_size; j++) {
      bool match = true;
      for (int pi = 0; pi < relevant_points.size(); pi++) {
        for (int pj = 0; pj < relevant_points[pi].size(); pj++) {
          if (complete_image[i + pi][j + relevant_points[pi][pj]] != '#') {
            match = false;
            break;
          }
        }
        if (!match) break;
      }
      if (match) count++;
    }
  }
  return count;
}

int main() {
  std::fstream file{"../input/day_20_input"};
  std::string line;
  Tile new_tile;
  std::unordered_map<int, Tile> tiles;
  std::unordered_map<std::string, std::vector<int>> edge_map;
  while (std::getline(file, line)) {
    line.erase(std::remove_if(std::begin(line), std::end(line),
                              [](const char c) { return !isprint(c); }),
               std::end(line));
    if (line == "") {
      new_tile.createEdges();
      tiles.insert({new_tile.id, new_tile});
      new_tile = Tile();
      continue;
    } else if (line.substr(0, 4) == "Tile") {
      new_tile.id = std::stoi(line.substr(5, line.size() - 5));
    } else {
      new_tile.image.push_back(line);
    }
  }

  if (new_tile.id != -1) {
    new_tile.createEdges();
    tiles.insert({new_tile.id, new_tile});
  }

  for (auto& [tile_id, tile] : tiles) {
    for (auto& edge : tile.edges) {
      if (edge_map.find(edge) == edge_map.end()) {
        edge_map.insert({edge, {tile_id}});
        std::string r_edge(std::rbegin(edge), std::rend(edge));
        edge_map.insert({r_edge, {tile_id}});
      } else {
        edge_map[edge].push_back(tile_id);
        std::string r_edge(std::rbegin(edge), std::rend(edge));
        edge_map[r_edge].push_back(tile_id);
      }
    }
  }

  // Create Border
  for (const auto& [edge, ids] : edge_map) {
    for (int i = 0; i < ids.size(); i++) {
      for (int j = i + 1; j < ids.size(); j++) {
        tiles[ids[i]].neighbour_ids.insert(ids[j]);
        tiles[ids[j]].neighbour_ids.insert(ids[i]);
      }
    }
  }

  std::unordered_set<int> corner_ids;
  for (auto& [id, tile] : tiles) {
    if (tile.neighbour_ids.size() == 2) {
      corner_ids.insert(id);
    }
  }

  bool found = false;
  long long count = 1;
  long long puzzle_size = std::sqrt(tiles.size());
  std::vector<std::vector<int>> grid(puzzle_size,
                                     std::vector<int>(puzzle_size, -1));
  // Using the corners as starting points, though the algorithm will work
  // starting form any tile
  for (const auto corner_id : corner_ids) {
    grid[0][0] = corner_id;
    tiles[corner_id].used = true;
    for (int i = 0; i < 4; ++i) {
      if (backtrack(puzzle_size, count, grid, edge_map, tiles)) {
        found = true;
        break;
      }
      tiles[corner_id].Rotate();
    }
    if (found) break;
    tiles[corner_id].Flip();
    for (int i = 0; i < 4; ++i) {
      if (backtrack(puzzle_size, count, grid, edge_map, tiles)) {
        found = true;
        break;
      }
      tiles[corner_id].Rotate();
    }
    if (found)
      break;
    else
      tiles[corner_id].used = false;
  }

  // Create complete image and remove edges
  std::vector<std::string> complete_image(
      (tiles[grid[0][0]].image.size() - 2) * (puzzle_size), "");
  for (int i = 0; i < puzzle_size; i++) {
    for (int j = 0; j < puzzle_size; j++) {
      tiles[grid[i][j]].applyRotationsAndFlipToImage();
      const auto& tile_img = tiles[grid[i][j]].image;
      for (int k = 0; k < (tile_img.size() - 2); k++) {
        complete_image[i * (tile_img.size() - 2) + k] +=
            tile_img[k + 1].substr(1, tile_img[k].size() - 2);
      }
    }
  }

  const std::vector<std::string> pattern = {
      "                  # ", "#    ##    ##    ###", " #  #  #  #  #  #   "};

  long long monster_size = 0;
  for (const auto& row : pattern) {
    for (const auto& ele : row) {
      if (ele == '#') {
        monster_size++;
      }
    }
  }

  long long n_monsters = 0;
  for (int i = 0; i < 2; i++) {    // flip
    for (int j = 0; j < 4; j++) {  // rotation
      n_monsters = CheckForPattern(complete_image, pattern);
      if (n_monsters > 0) break;
      RotateImage(complete_image);
    }
    if (n_monsters > 0) break;
    FlipImage(complete_image);
  }

  long long roughness = 0;
  for (int i = 0; i < complete_image.size(); i++) {
    for (int j = 0; j < complete_image[i].size(); j++) {
      if (complete_image[i][j] == '#') {
        roughness++;
      }
    }
  }

  roughness -= n_monsters * monster_size;
  std::cout << roughness << '\n';
  return roughness;
}
